### DISCUSSION TITLE
Why does Docker keeps containers with the "exit" status? Is there any relevant argument for that?

### DISCUSSION THREAD

**Discussion by BretFisher** [BretFisher/ama#190] [Q&A]
Because there may be some data in that container you want to keep. Docker doesn't assume that when a container exits that it should auto-destroy the changes that happened on the change layer that was created when you started the container. To have it auto remove, use `--rm` with `docker run`.

**Comment by subodh-dharma** [comment]
What type of data can you recover once container is exit. And how to access that?

**Comment by kawaii** [comment]
Do you think we'll ever see the extended form of `depends_on` whereby you can rely on a `healthcheck` to determine if it is time to raise a container or not, ported into the current Compose syntax (3.X), and active in the Swarm mode? It was a very useful feature as not all applications are resilient enough to just restart or idle until the database is ready again.

**Comment by BretFisher** [comment]
> What type of data can you recover once container is exit. And how to access that?

Assuming we're just talking about data in the container (not the data you may have saved in a docker volume or bind-mount), then your options are:

1. `docker container start` and then `docker container exec` or `docker container cp`
2. `docker container commit` saving container to a new image, then you can start a new container from that image, or do a `docker image save` to export the full file system.
3. I'm sure there are other easy ways

**Comment by BretFisher** [comment]
Hi @kawaii 

> Do you think we'll ever see the extended form of depends_on...

Highly unlikely because it's a general antipattern of a distributed system, legacy or not, and `depends_on` was designed for easier local development on a single host. That being said, the core libraries to implement this feature are open source, and largely determined by quality PR's.

What you're talking about is just the startup I take it, and it starts a debate about "what is startup when everything is up or down on the internet".

I'd need you to [explain how other waiting options/scripts](https://docs.docker.com/compose/startup-order/) or just letting docker restart/replace containers isn't a viable solution if you can't implement connection retries into code.
